### DISCUSSION TITLE
How can I manually run migrations?

### DISCUSSION THREAD

**Discussion by MiguelSOliveira** [simolus3/drift#1065] [Q&A] [answered]
I'm using a system where I do an account restore where I overwrite the "app.db" file with another file. I would like to not have to close the app in order for the migrations to be re-run. 

For example, I may overwrite a file with the schemaVersion of 6 with another with a schemaVersion of 3 and I would like for the migrations to run without the user closing the app.

Can I run them on demand?

**Comment by davidmartos96** [comment, marked as answer]
Migrations are run when the schema version of the database file in the device differs from the one that is passed when the db is opened. In the case of moor is through the getter in the AppDatabase file. 
You could create a constructor for the database that receives a specific version. Then you can simply close and reopen the database to trigger the migrations

**Reply by MiguelSOliveira** [reply]
I see there is a `close` method on the database, but I don't see an `open` or even `reopen` method. How would I reopen the database?

**Reply by davidmartos96** [reply]
Reopen is pretty much creating a new database object and run some query.
Alternatively you can open it without running a query by doing this:

```dart
final AppDatabase appDb = AppDatabase();
await appDb.executor.ensureOpen(appDb); 
```

**Reply by MiguelSOliveira** [reply]
Hmm, doesn't seem to work.. Is it actually possible to do this in runtime?

I'm closing the database, replacing the file, creating a new database and using Phoenix.rebirth (which is basically restarting the app by changing the key of the Root widget) and I get a black screen...

I was wondering if maybe the OS doesn't allow this at runtime, or maybe flutter even?

Here's how I'm doing it:

```
await DATABASE.close();
  final File file = File(dbLocation);
  await file.writeAsBytes(backupAsBytes);
  DATABASE = MoorDb(FlutterQueryExecutor.inDatabaseFolder(path: DATABASE_FILE));
  await DATABASE.executor.ensureOpen(DATABASE);
```

**Reply by davidmartos96** [reply]
It should be possible. Do you get an error in the console? The black screen is quite strange.

**Reply by MiguelSOliveira** [reply]
No error whatsoever.. It does run the "onUpgrade" tho... But I need to restart the app for it to open again after the black screen.

**Reply by davidmartos96** [reply]
If the onUpgrade gets called then you are on the right track regarding the migrations. Maybe the black screen error is related to the rebirth method.

**Reply by MiguelSOliveira** [reply]
Turns out it was being caught inside a try catch, it indeed looks related to moor:

`Unhandled Exception: Unhandled error DatabaseException(error database_closed)`

Any idea on what it could be?

I can confirm that all these print statements print before this error:

```
await DATABASE.close();
  print("Closing");
  final File file = File(dbLocation);
  await file.writeAsBytes(backupAsBytes);
  print("Writing bytes");
  DATABASE = MoorDb(FlutterQueryExecutor.inDatabaseFolder(path: DATABASE_FILE));
  print("Creating new DB");
  await DATABASE.executor.ensureOpen(DATABASE);
  print("It is open");
```

**Reply by davidmartos96** [reply]
It would be helpful to see the line where you get that error. Do you have the full error trace?

**Reply by davidmartos96** [reply]
You can do:

catch (e, st) {
    print(st);
    print(e);
}

**Reply by MiguelSOliveira** [reply]
```
[VERBOSE-2:ui_dart_state.cc(177)] Unhandled Exception: Unhandled error DatabaseException(error database_closed) occurred in Instance of 'XBloc'.
#0      SqfliteDatabaseMixin.checkNotClosed (package:sqflite_common/src/database_mixin.dart:282:7)
#1      SqfliteDatabaseExecutorMixin._rawQuery (package:sqflite_common/src/database_mixin.dart:125:8)
#2      SqfliteDatabaseExecutorMixin.rawQuery (package:sqflite_common/src/database_mixin.dart:120:12)
#3      _SqfliteExecutor.runSelect (package:moor_flutter/moor_flutter.dart:145:29)
#4      _ExecutorWithQueryDelegate.runSelect.<anonymous closure> (package:moor/src/runtime/executor/helpers/engines.dart:41:19)
#5      _ExecutorWithQueryDelegate._synchronized (package:moor/src/runtime/executor/helpers/engines.dart:25:26)
#6      _ExecutorWithQueryDelegate.runSelect (package:moor/src/runtime/executor/helpers/engines.dart:39:26)
#7      CustomSelectStatement._executeWithMappedArgs.<anonymous closure> (package:moor/src/runtime/query_builder/statements/sele<…>
```

**Reply by davidmartos96** [reply]
Yes, although that trace is missing the line in your project where it fails. Hard to tell otherwise 😅

**Reply by MiguelSOliveira** [reply]
It's just a `SELECT * FROM table;`. I think the problem is that the database is closed, no?

**Reply by davidmartos96** [reply]
You need to avoid running other queries while you are handling the database replacement.

**Reply by MiguelSOliveira** [reply]
I'm not tho :X We can see through the print statements that the table is already opened and replaced..

The particular error in my code is here: `database.table().getSingle();`

**Reply by MiguelSOliveira** [reply]
I think the problem might be in the way I'm reopening the database?

**Reply by davidmartos96** [reply]
Is possible that the instance of the database used there is the old one.
You can try printing the hashCode of the db after replacing and where it fails.

**Reply by davidmartos96** [reply]
I'd recommend restarting the app if possible with your requirements.
Or carefully handle the database instance after it has been recreated with get_it or with provider

**Reply by MiguelSOliveira** [reply]
You were totally right David, it was indeed the wrong hashCode, I'm now using getIt.

Sorry for the time it took for me to notice it, really appreciate your help man.

**Reply by davidmartos96** [reply]
No worries. Glad you got it working! ;)
